diesel/src/result.rs

```diff
@@ -94,6 +94,9 @@ pub enum DatabaseErrorKind {
     /// An example of a case where this would occur is if you attempted to send
     /// a query with more than 65000 bind parameters using PostgreSQL.
     UnableToSendCommand,
+    /// A lock could not be acquired, perhaps because the query was executed with
+    /// `FOR UPDATE NOWAIT` and the lock was already held.
+    LockNotAvailable,
```

Comment:
We've been very conservative about adding variants to this in the past. Can you provide some example use cases that need this variant?

Reply:
When using `FOR UPDATE NOWAIT` it's important to know whether the query failed due to the lock having been taken, or for some other reason. See the tests I added for a concrete example. I don't know how you'd use `NOWAIT` without that ability.

Reply:
It's worth noting that this is the only way to create a select statement that can fail at runtime in Diesel (other than network errors)

Reply:
That's not true, there is at least one other way a select could fail: timing out. In fact I use this in the test by setting a 1 second timeout to determine if `NOWAIT` is working correctly. (I deliberately set up a deadlock, and then if it times out instead of returning `LockNotAvailable` then it didn't work correctly).

I imagine there are also various encoding-related things that could go wrong in a select statement.

Reply:
Also if the schema as defined in rust doesn't match the database schema, that could also cause issues, or if permissions are not correct. There are many ways select could fail.

Reply:
I'm inclined to accept the variant, but I'd prefer that it be kept as a separate PR so we can keep the changes atomic, and to give that error a place to be discussed in isolation.

Reply:
What should I do about the test? I can't test `NOWAIT` without that error.

Reply:
Just test `is_err` for now

Reply:
OK, but that will always pass 😦

Reply:
Oh I see what you mean, it'd time out if it didn't fail immediately. Perhaps check that the error message contained an expected string?